File: omniclaw/voice/wake_phrase_matcher.py
"""Python mirror of WakePhraseMatcher.kt and assets/node/src/wake/phrase_matcher.js.

We deliberately keep the same constants and the same Damerau-Levenshtein
algorithm so a fixture that matches on Android also matches on Mac.

Algorithm (same on all three sides):

    1. Lowercase and strip non-alphanumeric punctuation.
    2. Tokenize on whitespace.
    3. Slide a window of len(target) tokens over candidate tokens; for each
       window compute per-token Damerau-Levenshtein distance, allow up to
       _max_token_edits_for(targetToken) edits per token AND up to
       MAX_TOTAL_EDITS overall.
    4. Match must start at a token boundary (so 'amber' does not fire 'ben').

Per-target-token edit limit:
    target token length <= 6: 0 edits (exact). Keeps "Ben"/"Sasha"/"Jarvis"/
        "Friday" all in exact-only mode - short wake words match too much
        ambient noise at 1 edit.
    target token length >= 7: MAX_TOKEN_EDITS edits.

The function returns True on first matching window.
"""
from __future__ import annotations

import re

MAX_TOKEN_EDITS = 1
MAX_TOTAL_EDITS = 2
STRICT_LENGTH_THRESHOLD = 6

_NORM_RE = re.compile(r"[^a-z0-9 ]")
_WS_RE = re.compile(r"\s+")


def _max_token_edits_for(token: str) -> int:
    return 0 if len(token) <= STRICT_LENGTH_THRESHOLD else MAX_TOKEN_EDITS


def matches(candidate: str | None, target: str) -> bool:
    if not candidate:
        return False
    target_tokens = [t for t in _normalize(target).split(" ") if t]
    if not target_tokens:
        return False
    cand_tokens = [t for t in _normalize(candidate).split(" ") if t]
    if len(cand_tokens) < len(target_tokens):
        return False

    for start in range(0, len(cand_tokens) - len(target_tokens) + 1):
        total = 0
        ok = True
        for i, tt in enumerate(target_tokens):
            edits = _damerau_levenshtein(tt, cand_tokens[start + i])
            if edits > _max_token_edits_for(tt):
                ok = False
                break
            total += edits
            if total > MAX_TOTAL_EDITS:
                ok = False
                break
        if ok:
            return True
    return False


def _normalize(s: str) -> str:
    s = s.lower()
    s = _NORM_RE.sub(" ", s)
    s = _WS_RE.sub(" ", s)
    return s.strip()


def _damerau_levenshtein(a: str, b: str) -> int:
    n, m = len(a), len(b)
    if n == 0:
        return m
    if m == 0:
        return n
    d = [[0] * (m + 1) for _ in range(n + 1)]
    for i in range(n + 1):
        d[i][0] = i
    for j in range(m + 1):
        d[0][j] = j
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            cost = 0 if a[i - 1] == b[j - 1] else 1
            d[i][j] = min(
                d[i - 1][j] + 1,
                d[i][j - 1] + 1,
                d[i - 1][j - 1] + cost,
            )
            if i > 1 and j > 1 and a[i - 1] == b[j - 2] and a[i - 2] == b[j - 1]:
                d[i][j] = min(d[i][j], d[i - 2][j - 2] + cost)
    return d[n][m]

File: android/app/src/main/java/com/ben/wake/WakePhraseMatcher.kt
package com.ben.wake

/**
 * Fuzzy matcher for the wake phrase.
 *
 * Algorithm:
 *   1. Lowercase + strip punctuation from both target and candidate.
 *   2. Tokenize on whitespace.
 *   3. Slide a window of `target.tokenCount` over candidate tokens; for each
 *      window compute the per-pair Damerau-Levenshtein distance, allow up to
 *      maxTokenEditsFor(targetToken) edits per pair AND at most
 *      MAX_TOTAL_EDITS overall.
 *   4. Match must START on a token boundary - protects against "amber" -> "ben".
 *
 * Per-target-token edit limit:
 *   - target token length <= 6: 0 edits (exact). Required because short
 *     wake words pick up too many false positives at 1 edit: "ben" matches
 *     "ban"/"bend"/"pen"; "sasha" matches "tasha"/"saska"/"sasher". The
 *     6-character cutoff keeps "Ben" / "Sasha" / "Jarvis" / "Friday" all
 *     in exact-match mode, which is what we want for a noisy ambient mic.
 *   - target token length >= 7: MAX_TOKEN_EDITS edits.
 *
 * Identical Kotlin / JS / Python implementations - keep them in sync.
 */
object WakePhraseMatcher {
    private const val MAX_TOKEN_EDITS = 1
    private const val MAX_TOTAL_EDITS = 2
    private const val STRICT_LENGTH_THRESHOLD = 6

    fun matches(candidate: String?, target: String): Boolean {
        if (candidate.isNullOrBlank()) return false
        val targetTokens = normalize(target).split(' ').filter { it.isNotEmpty() }
        if (targetTokens.isEmpty()) return false
        val candTokens = normalize(candidate).split(' ').filter { it.isNotEmpty() }
        if (candTokens.size < targetTokens.size) return false
        for (start in 0..(candTokens.size - targetTokens.size)) {
            var totalEdits = 0
            var ok = true
            for (i in targetTokens.indices) {
                val tt = targetTokens[i]
                val edits = damerauLevenshtein(tt, candTokens[start + i])
                if (edits > maxTokenEditsFor(tt)) { ok = false; break }
                totalEdits += edits
                if (totalEdits > MAX_TOTAL_EDITS) { ok = false; break }
            }
            if (ok) return true
        }
        return false
    }

    private fun maxTokenEditsFor(token: String): Int =
        if (token.length <= STRICT_LENGTH_THRESHOLD) 0 else MAX_TOKEN_EDITS

    private fun normalize(s: String): String =
        s.lowercase().replace(Regex("[^a-z0-9 ]"), " ").replace(Regex("\\s+"), " ").trim()

    /** Damerau-Levenshtein with adjacent-transposition. Small strings, O(n*m) is fine. */
    private fun damerauLevenshtein(a: String, b: String): Int {
        val n = a.length; val m = b.length
        if (n == 0) return m
        if (m == 0) return n
        val d = Array(n + 1) { IntArray(m + 1) }
        for (i in 0..n) d[i][0] = i
        for (j in 0..m) d[0][j] = j
        for (i in 1..n) {
            for (j in 1..m) {
                val cost = if (a[i - 1] == b[j - 1]) 0 else 1
                d[i][j] = minOf(
                    d[i - 1][j] + 1,
                    d[i][j - 1] + 1,
                    d[i - 1][j - 1] + cost,
                )
                if (i > 1 && j > 1 && a[i - 1] == b[j - 2] && a[i - 2] == b[j - 1]) {
                    d[i][j] = minOf(d[i][j], d[i - 2][j - 2] + cost)
                }
            }
        }
        return d[n][m]
    }
}
